Source/Dafny/Plugin.cs

```diff
@@ -5,44 +5,86 @@
 using Microsoft.Dafny.Plugins;
 
 namespace Microsoft.Dafny;
+using Dafny.Plugins;
+
+public interface Plugin {
+  public IEnumerable<Compiler> GetCompilers();
+  public IEnumerable<IRewriter> GetRewriters(ErrorReporter reporter);
+}
+
+public class ConfiguredPlugin : Plugin {
+  private readonly PluginConfiguration configuration;
+
+  public ConfiguredPlugin(PluginConfiguration configuration) {
+    this.configuration = configuration;
+  }
+
+  public IEnumerable<Compiler> GetCompilers() {
+    return configuration.GetCompilers();
+  }
+
+  public IEnumerable<IRewriter> GetRewriters(ErrorReporter reporter) {
+    return configuration.GetRewriters(reporter).Select(rewriter => new PluginRewriter(reporter, rewriter));
+  }
+}
 
 /// <summary>
 /// This class wraps an assembly and an extracted configuration from this assembly,
 /// The configuration provides the methods to parse command-line arguments and obtain Rewriters 
 /// </summary>
-public record Plugin(Assembly Assembly, string[] Args) {
-  public PluginConfiguration PluginConfiguration { get; } = LoadConfiguration(Assembly, Args);
+public class AssemblyPlugin : ConfiguredPlugin {
+  public AssemblyPlugin(Assembly assembly, string[] args) : base(LoadConfiguration(assembly, args)) {
+  }
 
   public static Plugin Load(string pluginPath, string[] args) {
-    return new Plugin(Assembly.LoadFrom(pluginPath), args);
+    return new AssemblyPlugin(Assembly.LoadFrom(pluginPath), args);
   }
 
   class AutomaticPluginConfiguration : PluginConfiguration {
-    public Func<ErrorReporter, Rewriter>[] Rewriters { get; init; }
-    public override Rewriter[] GetRewriters(ErrorReporter errorReporter) {
-      return Rewriters.Select(funcErrorReporterRewriter =>
-        funcErrorReporterRewriter(errorReporter)).ToArray();
+    public AutomaticPluginConfiguration(Assembly assembly) {
+      var types = assembly.GetTypes();
+
+      Rewriters = FindPluginComponents<Rewriter, Func<ErrorReporter, Rewriter>>(assembly, CreateRewriterFactory);
+      Compilers = FindPluginComponents<Compiler, Func<Compiler>>(assembly, CreateCompilerFactory);
+
+      // Checks about the plugin to be well-behaved.
+      if (Rewriters.Length == 0 && Compilers.Length == 0) {
```

Comment:
It could be possible that the assembly itself loads other assemblies, in which case there might not be Rewriter or Compiler directly in this assembly.
Please also check that there is no `PluginConfiguration` to match your error message.

Reply:
Hmm, no, I think this code path is only reached if there is no plugin configuration in the first place?

Reply:
oh ok that makes sense, because it wouldn't even try to list those.